### PR TITLE
Use ConcurrentDictionary's in DirectoryService

### PR DESCRIPTION
**Pull request by Spacetech**
I'm working on making the scanner / metadata refresh run multiple at once. One issue I ran into is `DirectoryService`'s cache dictionaries. Errors like this would show up since it's not thread safe:
```
System.NullReferenceException: Object reference not set to an instance of an object.
   at System.Collections.Generic.Dictionary`2.TryInsert(TKey key, TValue value, InsertionBehavior behavior)
   at System.Collections.Generic.Dictionary`2.set_Item(TKey key, TValue value)
   at MediaBrowser.Controller.Providers.DirectoryService.GetFile(String path) in C:\Programming\jellyfin\MediaBrowser.Controller\Providers\DirectoryService.cs:line 60
   at MediaBrowser.Providers.MediaInfo.FFProbeProvider.HasChanged(BaseItem item, IDirectoryService directoryService) in C:\Programming\jellyfin\MediaBrowser.Providers\MediaInfo\FFProbeProvider.cs:line 95
   at MediaBrowser.Providers.Manager.MetadataService`2.HasChanged(BaseItem item, IHasItemChangeMonitor changeMonitor, IDirectoryService directoryService) in C:\Programming\jellyfin\MediaBrowser.Providers\Manager\MetadataService.cs:line 910
```
Switching to use `ConcurrentDictionary` solves this.

**Changes**
- Use `ConcurrentDictionary`'s in `DirectoryService`

**Issues**
None

### CODE REVIEW

File: MediaBrowser.Controller/Providers/DirectoryService.cs [hunk 1]
```diff
@@ -51,36 +45,32 @@ public List<FileSystemMetadata> GetFiles(string path)
...
-                    _fileCache[path] = file;
-                }
-                else
-                {
-                    return null;
-                }
+            if (result == null)
+            {
+                // lets not store null results in the cache
+                _fileCache.TryRemove(path, out FileSystemMetadata removed);
```
Comment (crobibero) [inline MediaBrowser.Controller/Providers/DirectoryService.cs:57]:
```suggestion
                _fileCache.TryRemove(path, out _);
```
unused

File: MediaBrowser.Controller/Providers/DirectoryService.cs [hunk 2]
```diff
@@ -24,14 +25,7 @@ public DirectoryService(IFileSystem fileSystem)
...
         {
-            if (!_cache.TryGetValue(path, out FileSystemMetadata[] entries))
-            {
-                entries = _fileSystem.GetFileSystemEntries(path).ToArray();
-
-                _cache[path] = entries;
-            }
-
-            return entries;
+            return _cache.GetOrAdd(path, (p) => _fileSystem.GetFileSystemEntries(p).ToArray());
```
Comment (crobibero) [inline MediaBrowser.Controller/Providers/DirectoryService.cs:28]:
```suggestion
            return _cache.GetOrAdd(path, p => _fileSystem.GetFileSystemEntries(p).ToArray());
```

File: MediaBrowser.Controller/Providers/DirectoryService.cs [hunk 3]
```diff
@@ -51,36 +45,32 @@ public List<FileSystemMetadata> GetFiles(string path)
 
         public FileSystemMetadata GetFile(string path)
         {
-            if (!_fileCache.TryGetValue(path, out FileSystemMetadata file))
+            var result = _fileCache.GetOrAdd(path, (p) =>
```
Comment (crobibero) [inline MediaBrowser.Controller/Providers/DirectoryService.cs:48]:
```suggestion
            var result = _fileCache.GetOrAdd(path, p =>
```

File: MediaBrowser.Controller/Providers/DirectoryService.cs [hunk 4]
```diff
@@ -51,36 +45,32 @@ public List<FileSystemMetadata> GetFiles(string path)
...
             {
-                file = _fileSystem.GetFileInfo(path);
+                var file = _fileSystem.GetFileInfo(path);
+                return (file != null && file.Exists) ? file : null;
```
Comment (crobibero) [inline MediaBrowser.Controller/Providers/DirectoryService.cs:51]:
```suggestion
                return file != null && file.Exists ? file : null;
```

File: MediaBrowser.Controller/Providers/DirectoryService.cs [hunk 5]
```diff
@@ -51,36 +45,32 @@ public List<FileSystemMetadata> GetFiles(string path)
...
 
         public IReadOnlyList<string> GetFilePaths(string path, bool clearCache)
         {
-            if (clearCache || !_filePathCache.TryGetValue(path, out List<string> result))
+            if (clearCache)
             {
-                result = _fileSystem.GetFilePaths(path).ToList();
-
-                _filePathCache[path] = result;
+                _filePathCache.TryRemove(path, out List<string> removed);
```
Comment (crobibero) [inline MediaBrowser.Controller/Providers/DirectoryService.cs:70]:
```suggestion
                _filePathCache.TryRemove(path, out _);
```unused

File: MediaBrowser.Controller/Providers/DirectoryService.cs [hunk 6]
```diff
@@ -51,36 +45,32 @@ public List<FileSystemMetadata> GetFiles(string path)
...
             }
 
-            return result;
+            return _filePathCache.GetOrAdd(path, (p) => _fileSystem.GetFilePaths(path).ToList());
```
Comment (crobibero) [inline MediaBrowser.Controller/Providers/DirectoryService.cs:73]:
```suggestion
            return _filePathCache.GetOrAdd(path, p => _fileSystem.GetFilePaths(path).ToList());
```

File: MediaBrowser.Controller/Providers/DirectoryService.cs [hunk 7]
```diff
@@ -51,36 +45,32 @@ public List<FileSystemMetadata> GetFiles(string path)
 
         public FileSystemMetadata GetFile(string path)
         {
-            if (!_fileCache.TryGetValue(path, out FileSystemMetadata file))
+            var result = _fileCache.GetOrAdd(path, p =>
             {
-                file = _fileSystem.GetFileInfo(path);
+                var file = _fileSystem.GetFileInfo(path);
```
Comment (crobibero) [inline MediaBrowser.Controller/Providers/DirectoryService.cs:50]:
```suggestion
                var file = _fileSystem.GetFileInfo(p);
```

**Review by crobibero — APPROVED** [review]

**Review by barronpm — APPROVED** [review]

**Pull request merged**